vm/devices/storage/disk_nvme/nvme_driver/src/driver.rs

```diff
@@ -432,6 +495,195 @@ impl<T: DeviceBacking> NvmeDriver<T> {
+
+        let dma_buffer = worker.device.host_allocator();
+        // Restore the admin queue pair.
+        let admin = saved_state
+            .admin
+            .as_ref()
+            .map(|a| {
+                // Restore memory block for admin queue pair.
+                let mem_block = dma_buffer
+                    .attach_dma_buffer(a.mem_len, a.pfns.as_slice())
+                    .expect("unable to restore mem block");
+                QueuePair::restore(driver.clone(), interrupt0, registers.clone(), mem_block, a)
+                    .unwrap()
+            })
+            .unwrap();
+
+        let admin = worker.admin.insert(admin);
+
+        // Spawn a task to handle asynchronous events.
+        let async_event_task = this.driver.spawn("nvme_async_event", {
+            let admin = admin.issuer().clone();
+            let rescan_event = this.rescan_event.clone();
+            async move {
+                if let Err(err) = handle_asynchronous_events(&admin, &rescan_event).await {
+                    tracing::error!(
+                        error = err.as_ref() as &dyn std::error::Error,
+                        "asynchronous event failure, not processing any more"
+                    );
+                }
+            }
+        });
+
+        let state = WorkerState {
+            qsize: saved_state.qsize,
+            async_event_task,
+            max_io_queues: saved_state.max_io_queues,
+        };
+
+        this.admin = Some(admin.issuer().clone());
+
+        // Restore I/O queues.
+        // Interrupt vector 0 is shared between Admin queue and I/O queue #1.
+        let mut ioq: Vec<IoQueue> = Vec::new();
+        for q_state in &saved_state.io {
+            let interrupt = worker
+                .device
+                .map_interrupt(q_state.msix, q_state.cpu)
+                .context("failed to map interrupt")?;
+
+            let mem_block =
+                dma_buffer.attach_dma_buffer(q_state.mem_len, q_state.pfns.as_slice())?;
+            let q = IoQueue::restore(
+                driver.clone(),
+                interrupt,
+                registers.clone(),
+                mem_block,
+                q_state,
+            )
+            .unwrap();
+
+            let issuer = IoIssuer {
+                issuer: q.queue.issuer().clone(),
+                cpu: q_state.cpu,
+            };
+            this.io_issuers.per_cpu[q_state.cpu as usize]
+                .set(issuer)
+                .unwrap();
+
+            ioq.push(q);
+        }
+        worker.io = ioq;
+
+        // Restore namespace(s).
+        for ns in &saved_state.namespaces {
+            // TODO: Current approach is to re-query namespace data after servicing
+            // and this array will be empty. Once we confirm that we can process
+            // namespace change notification AEN, the restore code will be re-added.
+            this.namespaces.push(Arc::new(Namespace::restore(
```

Comment:
If we're not going to do this yet, then let's remove this `namespaces` field and put `#[allow(dead_code)]` on `Namespace::restore`.

Reply:
ok

Reply:
Added `#[allow(dead_code)]` to `Namespace::save` instead - save is not called anywhere in the code and restore will be skipped because the vector is empty.